# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

/opt/anaconda3/envs/myenv/lib/python3.10/site-packages/threadpoolctl.py:1214: RuntimeWarning: 
Found Intel OpenMP ('libiomp') and LLVM OpenMP ('libomp') loaded at
the same time. Both libraries are known to be incompatible and this
can cause random crashes or deadlocks on Linux when loaded in the
same Python program.
Using threadpoolctl may cause crashes or deadlocks. For more
information and possible workarounds, please see
    https://github.com/joblib/threadpoolctl/blob/master/multiple_openmp.md

  warnings.warn(msg, RuntimeWarning)


## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

*Replace the text in this markdown cell with your predictions, and be sure to provide justification for your guess.*

## Split the Data into Training and Testing Sets

In [12]:
data.shape

(4601, 58)

In [13]:
data.info

<bound method DataFrame.info of       word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0               0.00               0.64           0.64           0.0   
1               0.21               0.28           0.50           0.0   
2               0.06               0.00           0.71           0.0   
3               0.00               0.00           0.00           0.0   
4               0.00               0.00           0.00           0.0   
...              ...                ...            ...           ...   
4596            0.31               0.00           0.62           0.0   
4597            0.00               0.00           0.00           0.0   
4598            0.30               0.00           0.30           0.0   
4599            0.96               0.00           0.00           0.0   
4600            0.00               0.00           0.65           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0              0.32       

In [14]:
data.sample(10)

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
3910            0.00               0.00           0.30           0.0   
2826            0.00               0.00           0.00           0.0   
1943            0.00               0.00           0.00           0.0   
949             0.00               0.00           0.00           0.0   
2564            0.00               0.00           2.50           0.0   
634             0.00               0.35           0.71           0.0   
2051            0.00               0.00           0.00           0.0   
324             0.00               0.00           0.00           0.0   
1846            0.00               0.00           0.79           0.0   
92              0.59               0.44           0.29           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
3910           0.61            0.00              0.00                0.00   
2826           0.00            0.00              0.00                0.00   
1943           0.00            0.00              0.00                0.00   
949            0.20            0.00              0.00                0.00   
2564           0.00            0.00              0.00                0.00   
634            0.35            0.00              0.17                0.00   
2051           0.00            0.00              0.00                0.00   
324            0.00            0.00              0.00                0.67   
1846           0.00            0.00              0.00                0.00   
92             0.14            0.03              0.03                0.14   

      word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
3910             0.00            0.00  ...        0.301        0.043   
2826             0.00            0.00  ...        0.000        0.000   
1943             0.00            0.00  ...        0.000        0.165   
949              0.00            0.00  ...        0.000        0.204   
2564             0.00            0.00  ...        0.000        0.000   
634              0.00            0.53  ...        0.000        0.027   
2051             0.00            0.00  ...        0.000        0.000   
324              0.00            0.67  ...        0.170        0.511   
1846             0.00            0.00  ...        0.000        0.237   
92               0.56            0.67  ...        0.000        0.075   

      char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
3910        0.043        0.000        0.086        0.000   
2826        0.000        0.344        0.000        0.000   
1943        0.000        0.165        0.000        0.000   
949         0.000        0.034        0.000        0.000   
2564        0.000        0.000        0.000        0.000   
634         0.000        0.167        0.027        0.055   
2051        0.000        0.000        0.000        0.000   
324         0.085        0.511        0.000        0.000   
1846        0.000        0.000        0.000        0.059   
92          0.000        0.613        0.532        0.137   

      capital_run_length_average  capital_run_length_longest  \
3910                       2.161                          19   
2826                       3.250                          17   
1943                       1.666                           7   
949                        2.588                          15   
2564                       1.000                           1   
634                        7.527                         149   
2051                       1.000                           1   
324                        4.617                          27   
1846                       2.510                          12   
92                         7.300                         763   

      capital_run_length_total  spam  
3910                       227     0  
2826                        52     0  
1943                        45     0  
949                        277     1  
2564       

In [8]:
data.columns 

Index(['word_freq_make', 'word_freq_address', 'word_freq_all', 'word_freq_3d',
       'word_freq_our', 'word_freq_over', 'word_freq_remove',
       'word_freq_internet', 'word_freq_order', 'word_freq_mail',
       'word_freq_receive', 'word_freq_will', 'word_freq_people',
       'word_freq_report', 'word_freq_addresses', 'word_freq_free',
       'word_freq_business', 'word_freq_email', 'word_freq_you',
       'word_freq_credit', 'word_freq_your', 'word_freq_font', 'word_freq_000',
       'word_freq_money', 'word_freq_hp', 'word_freq_hpl', 'word_freq_george',
       'word_freq_650', 'word_freq_lab', 'word_freq_labs', 'word_freq_telnet',
       'word_freq_857', 'word_freq_data', 'word_freq_415', 'word_freq_85',
       'word_freq_technology', 'word_freq_1999', 'word_freq_parts',
       'word_freq_pm', 'word_freq_direct', 'word_freq_cs', 'word_freq_meeting',
       'word_freq_original', 'word_freq_project', 'word_freq_re',
       'word_freq_edu', 'word_freq_table', 'word_freq_conference',


In [15]:
# Create the labels set `y` and features DataFrame `X`
y = data['spam']
X = data.drop('spam', axis='columns')


In [16]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [17]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y)

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [18]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()

In [19]:
# Fit the Standard Scaler with the training data
scaler = scaler.fit(X_train)

In [20]:
# Scale the training data
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [21]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression


In [22]:
# Make and save testing predictions with the saved logistic regression model using the test data
lr_model = LogisticRegression(random_state=1)
lr_model.fit(X_train_scaled, y_train)

# Review the predictions
lr_test_predictions = lr_model.predict(X_test_scaled)
lr_test_predictions

array([1, 0, 0, ..., 1, 0, 1])

In [23]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, lr_test_predictions)

0.9235447437011295

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [24]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier


In [25]:
# Make and save testing predictions with the saved logistic regression model using the test data
rfc_model = RandomForestClassifier(random_state=1)
rfc_model.fit(X_train_scaled, y_train)

# Review the predictions
rfc_test_predictions = rfc_model.predict(X_test_scaled)

In [26]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, rfc_test_predictions)

0.945264986967854

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

### Conclusion
#1. Logistic Regression achieved an accuracy of 0.922.
#2. Random Forest Classifier reached approximately 0.945 accuracy.
#3. Random Forest was more accurate, aligning with my prediction.
